pkg/summarizer/flakiness.go

```diff
@@ -0,0 +1,286 @@
+
+// Keep in mind that flakiness is measured as out of 100, i.e. 23 not .23
+var DEFAULT_FLAKINESS = 50.0
+var MIN_RUNS = 0
+var HEALTHY_RANGE = map[int]string{
+	20: "red",
+	10: "purple",
+	3:  "orange",
+	0:  "green",
+}
+
+type Result struct {
+	name             string
+	passed           int
+	failed           int
+	flakyCount       int
+	averageFlakiness float64
+	failedInfraCount int
+	infraFailures    map[string]int
+}
+
+// Temporary structs while I decide what will go into summary.proto
+type Healthiness struct {
+	startDate        int
+	endDate          int
+	tests            []TestInfo
+	totalTests       int
+	totalJailedTests int
+	averageFlakiness float64
+	flakyBuckets     []FlakyBucket
+	infraIssues      map[string]int
+	totalConfigs     int
+}
+
+type TestInfo struct {
+	name               string
+	env                string
+	totalRuns          int
+	totalRunsWithInfra int
+	passedRuns         int
+	failedRuns         int
+	failedInfraRuns    int
+	flakyRuns          int
+	flakiness          float64
+	infraInfo          string
+}
+
+type FlakyBucket struct {
+	threshold float64
+	tests     int
+}
+
+// For sorting, primarily intended for map[string]int as implemented below
+type IntString struct {
+	i int
+	s string
+}
+
+func CalculateHealthiness(grid *state.Grid, startTime int, endTime int, tab string) Healthiness {
+	results := parseGrid(grid, startTime, endTime)
+	return analyzeFlakinessFromResults(results, startTime, endTime, tab)
+}
+
+func parseGrid(grid *state.Grid, startTime int, endTime int) []Result {
+	// Get the relevant data for flakiness from each Grid (which represents
+	// a dashboard tab) as a list of Result structs
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	results := make([]Result, 0)
+
+	for _, test := range grid.Rows {
+		// each row in Rows will have a field results
+		// results is run length encoded
+		// e.g. (encoded -> decoded equivalent):
+		// [0, 3, 5, 4] -> [0, 0, 0, 5, 5, 5, 5]
+		// decoded int values correspond to Row.Result enum
+		var resultCounts Result
+		resultCounts.infraFailures = make(map[string]int)
+		i := -1
+		for testResult := range result.Iter(ctx, test.Results) {
+			i += 1
+			if !isWithinTimeFrame(grid.Columns[i], startTime, endTime) {
+				continue
+			}
+			switch rowResult := result.Coalesce(testResult, result.IgnoreRunning); rowResult {
+			case state.Row_NO_RESULT:
+				continue
+			case state.Row_FAIL:
+				categorizeFailure(&resultCounts, test.Messages[i])
+			case state.Row_PASS:
+				resultCounts.passed += 1
+			case state.Row_FLAKY:
+				getValueOfFlakyResult(&resultCounts)
+			}
+		}
+		if resultCounts.failed > 0 || resultCounts.passed > 0 || resultCounts.flakyCount > 0 {
+			resultCounts.name = test.Name
+			results = append(results, resultCounts)
+		}
+
+	}
+	return results
+}
+
+func analyzeFlakinessFromResults(results []Result, startTime int, endTime int, tab string) Healthiness {
```

Comment:
Is this just to be a framework to enable other types of flake analysis?

Reply:
I feel like we should create a `FlakinessAnalyser` interface, and this naiveFlakiness can be the first implementation. This will make future expansion easier.

Reply:
Done